File: 2035-count-sub-islands/count-sub-islands.cpp
class Solution {
public:
    vector<int> x = {0, 0, 1, -1};
    vector<int> y = {1, -1, 0, 0};
    bool island = false;
    void dfs(int i, int j, vector<vector<int>>& grid1, vector<vector<int>>& grid2, int m, int n) {
        if (grid1[i][j] != grid2[i][j]) {
            island = false;
            return;
        }

        grid2[i][j] = grid1[i][j] = 0;

        for (int k = 0; k < 4; k++) {
            int newX = i + x[k];
            int newY = j + y[k];
            if (newX >= 0 && newX < m && newY >= 0 && newY < n && grid2[newX][newY] == 1) 
                dfs(newX, newY, grid1, grid2, m, n);
        }
    }
    int countSubIslands(vector<vector<int>>& grid1, vector<vector<int>>& grid2) {
        int count = 0;
        int m = grid1.size();
        int n = grid1[0].size();

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid2[i][j] == 1) {
                    island = true;
                    dfs(i, j, grid1, grid2, m, n);
                    if (island) count++;
                }
            }
        }

        return count;
    }
};

File: 2035-count-sub-islands/count-sub-islands.java
class Solution {
    int[] dx = new int[]{-1, 0, 1, 0};
    int[] dy = new int[]{0, 1, 0, -1};

    public int countSubIslands(int[][] grid1, int[][] grid2) {
        int m = grid1.length, n = grid1[0].length;
        int res = 0;
        boolean[][] visited = new boolean[m][n];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid2[i][j] == 0 || visited[i][j]) continue;

                res += dfs(i, j, m, n, grid1, grid2, visited) == true ? 1 : 0;
            }
        }

        return res;
    }

    private boolean dfs(int i, int j, int m, int n, int[][] grid1, int[][] grid2, boolean[][] visited) {
        visited[i][j] = true;
        boolean ok = grid1[i][j] == 1;

        for (int k = 0; k < 4; k++) {
            int x = i + dx[k];
            int y = j + dy[k];

            if (!isInside(x, y, m, n) || visited[x][y] || grid2[x][y] == 0) continue;

            ok &= dfs(x, y, m, n, grid1, grid2, visited);
        }

        return ok;
    }

    private boolean isInside(int x, int y, int m, int n) {
        return x >= 0 && x < m && y >= 0 && y < n;
    }
}